activerecord/lib/active_record/associations/has_one_association.rb

```diff
@@ -56,12 +56,18 @@ def replace(record, save = true)
 
           return target unless load_target || record
 
-          assigning_another_record = target != record
+          record_foreign_key_changed = record && record.changes_to_save.include?(reflection.foreign_key)
+          assigning_another_record = target != record || record_foreign_key_changed
```

Comment:
Isn't checking if the record_foreign_key_changed enough to fix this bug?
`target != record` would return `true` if it has been previously assigned.